### DISCUSSION TITLE
Compiling DB4S on Windows

### DISCUSSION THREAD

**Discussion by chrisjlocke** [sqlitebrowser/sqlitebrowser#3150] [General]
@longnguyen2004 - You've built this on Windows, right?  Can I pick your brains?
I've installed A Win 10 VM and downloaded 'the bits'

![image](https://user-images.githubusercontent.com/7872218/185809367-42a14271-f761-417f-8533-d74a15c3a31a.png)

Using VS 2017 to compile 'ALL_BUILD' works fine - no errors.

![image](https://user-images.githubusercontent.com/7872218/185809426-deceda59-e9b7-436f-9261-babb197d0f01.png)

![image](https://user-images.githubusercontent.com/7872218/185809444-0b4987e8-23f7-413c-9f51-1a11bf928331.png)

But the .exe produced is 16 MB, not ~ 6 MB.  Also running it doesn't work.

![image](https://user-images.githubusercontent.com/7872218/185809489-2bd85002-a2cb-4825-9f3e-962e77b2bbf6.png)

(this was copying the .exe to a directory containing the current nightly)

Building the 'INSTALL' project didn't work

![image](https://user-images.githubusercontent.com/7872218/185809552-dbc5728c-4786-4896-b49c-63f51d0c1da0.png)

I'll add that the step 'build project file' generated an error, but 'sticking my head in the sand and ignoring it' seemed to work, and it created a solution file anyway?

![image](https://user-images.githubusercontent.com/7872218/185809615-81276944-9d89-4b2f-b087-2060286b2919.png)

I've been following the Win 64 build guide and currently got up to the install part (https://github.com/sqlitebrowser/sqlitebrowser/wiki/Win64-setup-%E2%80%94-Step-13-%E2%80%94-Install-DB4S-locally).  I assumed I didn't need to actually follow this - I could just grab the .exe and plop it in the existing DB4S directory?

For the avoidance of doubt, I have no idea what I'm doing.  I'm OK with computers, but only in so far as working in my comfort zone ... tinkering with cmake and knowing what the wacky errors mean is outside that!

When building 'ALL-BUILD' it did report this:

3>qhexedit.vcxproj -> C:\git_repos\sqlitebrowser\libs\qhexedit\Debug\qhexedit.lib
4>------ Rebuild All started: Project: qscintilla2, Configuration: Debug x64 ------
4>C:/git_repos/sqlitebrowser/libs/qscintilla/Qt4Qt5/SciAccessibility.h(0): Note: No relevant classes found. No output generated.
4>Generating Qsci/moc_qsciscintilla.cpp

Also, when installing Qt, I couldn't see a 2017 flavour, so selected both 2015 and 2019...

![image](https://user-images.githubusercontent.com/7872218/185810226-145561d1-80a5-4726-8f38-8e3eb2c5f7bc.png)

Any help you can throw my way would be appreciated. :)

**Comment by justinclift** [comment]
As a thought, we use [Wix Toolset](https://wixtoolset.org) instead of NSIS these days for the packaging step.  And yeah, we need to update the wiki page to mention that too. :smile:

Looking on the Windows nightly build VM, it has Qt 5.12.8 installed.  That's *probably* where the MSVC 2017 compatibility is.

That being said, the macOS build VM has Qt 5.15.2 installed on it (just checked).  So there shouldn't be any DB4S source code compatibility problems with your selection of Qt 5.15.2 on Windows.  In theory. :wink:

**Comment by justinclift** [comment]
Hmmm, thinking about it more.  Perhaps the right approach for Win here would be to try installing VS 2019 instead?  That'd then probably work with the Qt 5.15.2 piece.

**Reply by longnguyen2004** [reply]
Yes, it's always good to use the most recent version (actually I'm using 2022 now, but it's binary compatible with 2019 so there's no problem)

**Comment by longnguyen2004** [comment]
16MB is due to you making a Debug build. If I switch to Release it drops to 5.5MB, like usual
![image](https://user-images.githubusercontent.com/15729831/185884373-b05bd5ff-aec5-4940-86b2-43645d78f528.png)

After copying the required DLLs to the executable folder, it works (about the missing icons, maybe I missed a plugin somewhere, maybe something's wrong with Qt itself, I don't know enough about it to fix it on my own)
![image](https://user-images.githubusercontent.com/15729831/185886564-2502c420-c78b-4380-95b1-f41b07a8f0e9.png)
![image](https://user-images.githubusercontent.com/15729831/185886677-8ee127a9-c272-4976-b785-4339c47f87af.png)

So, assuming Qt is in `C:\dev\Qt\5.15.2\msvc2019_64`, and SQLite is in `C:\dev\SQLite-Win64`, all you need to do is run
`cmake -DCMAKE_SYSTEM_PREFIX_PATH="C:\dev\Qt\5.15.2\msvc2019_64;C:\dev\SQLite-Win64" -B build`
then go to the `build` directory, and build the solution like usual.

If everything goes this, this should be the output you see
```
C:\git_repos\sqlitebrowser>cmake -DCMAKE_SYSTEM_PREFIX_PATH="C:\dev\Qt\5.15.2\msvc2019_64;C:\dev\SQLite-Win64" -B build
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.22000.
-- Could NOT find QScintilla (missing: QSCINTILLA_LIBRARY QSCINTILLA_INCLUDE_DIR) (Required is at least version "2.8.10")
-- Found SQLite3: C:/dev/SQLite-Win64 (found version "3.39.2")
-- Configuring done
-- Generating done
-- Build files have been written to: C:/git_repos/sqlitebrowser/build
```

**Reply by chrisjlocke** [reply]
Thanks for your help.  I installed VS 2019, so happy to try VS 2022 too.

> all you need to do is run

So going by the 'official' instructions, this is on [page 11](https://github.com/sqlitebrowser/sqlitebrowser/wiki/Win64-setup-%E2%80%94-Step-11-%E2%80%94-Generate-MSVC-solution-file). Do I still run the 'cmake -G "Visual Studio 12 2013 Win64" -Wno-dev' step, or just run yours?

Building with VS 2019 (but not running with the above cmake command yet) still didn't make a working .exe, but it did throw up some errors instead of just dying.  It moaned about missing qt5 DLLs - eg, Qt5networkd.dll.  I'll see what happens after stabbing about with the above cmake...

Again, thanks for your help 'n suggestions. :)

**Reply by longnguyen2004** [reply]
Yep that's step 11, try to use mine to see if there's any difference
Also, you need to copy the Qt's dll to the build folder, or add the Qt `bin` folder to PATH, so it can find the dll

**Reply by chrisjlocke** [reply]
Well, 'yay!' and 'bah!'

yay in that VS 2019 compiled it all OK with no errors.  Your build instructions helped.
Bah in that I got the same error.
The 0xc000007b error relates to a DLL not found, which is promising, but it doesn't say which one.  SysInternals Process Monitor isn't much help either.

![image](https://user-images.githubusercontent.com/7872218/185993663-a8f740bc-4166-4ae7-aa66-f9e0f4c314d9.png)

Well, I'll go through the steps again in case I goofed along the way...

**Reply by longnguyen2004** [reply]
That's really weird since 0xc000007b is [`STATUS_INVALID_IMAGE_FORMAT`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55#:~:text=0xC000007B,STATUS_INVALID_IMAGE_FORMAT), which I don't know how it could be related to this. Anyway, I saw some `NAME NOT FOUND` for `sqlite3.dll`, so maybe just have a check on that?

**Comment by longnguyen2004** [comment]
Re: SQLCipher
I can't find an amalgamated version of SQLCipher that's up-to-date, so I guess I'll have to generate it myself.

**Comment by justinclift** [comment]
Yeah.  Building SQLCipher yourself shouldn't be too hard.  This is the command we use for it in our nightly build script:

https://github.com/sqlitebrowser/sqlitebrowser/blob/a78dd0292f7e1149b4e81a3987531aa67098b92c/installer/windows/nightly_build_scripts/win64build.bat#L95-L97

That gets run in a clone of the SQLCipher GitHub source repo (https://github.com/sqlcipher/sqlcipher).  Earlier in the build script is this, though ignore the branch name / version number in the comment as that was from when the script was originally created:

https://github.com/sqlitebrowser/sqlitebrowser/blob/a78dd0292f7e1149b4e81a3987531aa67098b92c/installer/windows/nightly_build_scripts/win64build.bat#L26-L36

Note that I'm not 100% sure if the very latest release works.  I have a faint memory that we might have needed to keep back a minor version or so, for compiling with VS 2017 on the Windows build server.  Not super sure any more. :wink:

**Reply by longnguyen2004** [reply]
It's just that I'll need to have Tcl to build it, which is yet another thing I'll never use :)

**Reply by longnguyen2004** [reply]
And since SQLCipher on Windows is different (as usual), I'll probably have to fix `FindSQLCipher.cmake` for proper detection.

**Reply by longnguyen2004** [reply]
Also, for easy building dependencies from source, I'd like to introduce you to [vcpkg](https://github.com/microsoft/vcpkg). It lines up nicely with the way `FindSQLCipher.cmake` works.

**Reply by chrisjlocke** [reply]
"In order to use vcpkg with CMake outside of an IDE, you can use the toolchain file:"
/eyes glaze over....

They need more noddy instructions! ;)

**Comment by longnguyen2004** [comment]
How is stuff turning out?

**Reply by chrisjlocke** [reply]
Well, it started badly, tailed off in the middle, and the less said about the ending, the better. 😁😂
